src/Filters/remez_fir.jl

```
@@ -0,0 +1,825 @@
+            end
+        end
+    else
+        if !nodd
+            for j = 1 : ngrid
+                change = sinpi(grid[j])
+                des[j] = des[j] / change
+                wt[j]  = wt[j]  * change
+            end
+        else
+            for j = 1 : ngrid
+                change = sinpi(2grid[j])
+                des[j] = des[j] / change
+                wt[j]  = wt[j]  * change
+            end
+        end
+    end
+        
+    iext = zeros(Int64, dimsize)   # indices of extremals
+    x = zeros(Float64, dimsize)
+    y = zeros(Float64, dimsize)
+
+    for j = 1 : nfcns
+        iext[j] = Int64(floor((j-1)*temp)) + 1
+    end
+    iext[nfcns+1] = ngrid
+
+    dev = 0.0     # deviation from the desired function, 
+                  # that is, the amount of "ripple" on the extremal set
+    devl = -1.0   # deviation on last iteration
+    nz  = nfcns+1
+    nzz = nfcns+2
+    niter = 0
+    ad = zeros(Float64, nz)
+    
+    jet = ((nfcns-1) ÷ 15) + 1
+
+    while true
+    
+        #
+        # Start next iteration
+        #        
+      @label L100
+        iext[nzz] = ngrid + 1
+        niter += 1
+        if niter > maxiter
+            break
+        end
+        
+        x[1:nz] = cospi.(2grid[iext[1:nz]])
+        #for j = 1 : nz
+        #    printfmtln("  j={}: iext[j]={} grid[iext[j]]={} x[j]={}", j, iext[j], grid[iext[j]], x[j])
+        #end
+        
+        for j = 1 : nz
+            ad[j] = lagrange_interp(j, nz, jet, x)
+        end
+
+        dnum = 0.0
+        dden = 0.0
+        k = 1
+        for j = 1 : nz
+            l = iext[j]
+            dnum += ad[j] * des[l]
+            dden += k * ad[j] / wt[l]
+            k = -k
+        end
+        dev = dnum / dden
+        # printfmtln("DEVIATION = {}", dev)
+
+        y = 0*x
+        nu, dev = initialize_y(dev, nz, iext, des, wt, y)
+        
+        if dev <= devl
+            # finished 
+            return -1  # error - deviation should always increase
```

I think we should either throw an exception or return `nothing`. Where I'd prefer the former.